[styles] Add London Overground icon

charlie2clarke opened the pull request:
Part of #3419

**This is a WIP** as the icon isn't being shown yet.

The only OSM tag available to distinguish the London Overground service is `operator=` (see below)
<img src="https://github.com/organicmaps/organicmaps/assets/58811769/a114a2ca-8be9-4372-961f-4e4b73af52f6" width="30%" height="30%" />

I think the fact 'London Overground' has spaces in it might be tripping something up, though I've tried wrapping it in quotes in `Subways.mapcss` and it made no difference.

_Note: I will add this as a POI type once the icon is sorted._

RedAuburn commented:
icons look good 👍️

thinking about adding future overground rail, maybe it'd be best to add the type as `railway|station|heavy|london_overground`? 
then we have the main types of `subway`, `light` & `heavy` (see [Passenger Rail Terminology#Heavy Rail](https://en.wikipedia.org/wiki/Passenger_rail_terminology#Heavy_rail).

if you haven't regenerated a map file yet, there will be no `railway|station|overground|london` type in the data yet, so the icons won't be displayed even if the `subways.mapcss` is correct

charlie2clarke commented:
> maybe it'd be best to add the type as railway|station|heavy|london_overground

Sounds sensible, though isn't the Overground a type of subway? Thanks for sharing the terminology I didn't know what the diff was.

I thought I had regenerated the maps, but maybe I haven't done it right, where have I gone wrong in my steps:
- Follow the steps to add an icon outlined [in docs](https://github.com/organicmaps/organicmaps/blob/master/docs/STYLES.md#how-to-add-a-new-icon)
- Build generator tool
- Moved my `UK_England_Greater London.mwm` file from `data/240326` into `data/` 
- Ran: `$ ../omim-build-release/generator_tool --generate_index=true --output="UK_England_Greater London"`
- Replaced `data/240326/UK_England_Greater London.mwm` file with the output .mwm file from the above
- I also tried deleting each .mwm file and re-downloading but it made no difference

RedAuburn commented:
That's the correct process, it does sound like the type isn't being correctly matched for some reason

what type is displayed when you click on a station in the desktop app?
here's an example of a DLR station, you can see its type has been matched to `railway-station-light_rail-london`
![image](https://github.com/organicmaps/organicmaps/assets/26939824/ee24d868-2c6d-4d25-aac6-68646941401b)

> though isn't the Overground a type of subway?

[Wikipedia says](https://en.wikipedia.org/wiki/London_Overground) it's a "suburban rail network", I think for OM's purposes it's only really worth differentiating light/heavy/subway/tram though, as for the user a "suburban rail" feels and acts generally the same as an "inter-city rail" etc :)

charlie2clarke commented:
![image](https://github.com/organicmaps/organicmaps/assets/58811769/c9c92553-8035-4a49-99cf-cfdf7faa58a0)

It doesn't appear to get matched. On a side note - how is it that you can view a lot more info on the 'Place Page' dialogue than me even when I am using a debug build?

Do you know where the code is that matches Subways.mapcss to OSM tags?

RedAuburn commented:
> On a side note - how is it that you can view a lot more info on the 'Place Page' dialogue than me even when I am using a debug build?

Looks like i haven't rebuilt desktop in a while, it's because https://github.com/organicmaps/organicmaps/pull/7505 was merged. Now you need to enable developer mode in help (top right) -> preferences -> developer mode (tick) 👍️

> Do you know where the code is that matches Subways.mapcss to OSM tags?

it's just the standard matching code that's used for all types: https://github.com/organicmaps/organicmaps/blob/c214e44ecf99c512b1ad4207f374235c2ff35346/generator/osm2type.cpp#L396

the difference for subways etc is that the `city=*` is added to the element before it's converted to a type at https://github.com/organicmaps/organicmaps/blob/c214e44ecf99c512b1ad4207f374235c2ff35346/generator/osm2type.cpp#L683

RedAuburn commented:
maybe try processing operator from https://github.com/organicmaps/organicmaps/blob/c214e44ecf99c512b1ad4207f374235c2ff35346/generator/osm2type.cpp#L1124-L1147

i don't _think_ operator is discarded, but it might be

pastk commented:
> Do you know where the code is that matches Subways.mapcss to OSM tags?

First of all you need to add new types/mapping into `data/mapcss-mapping.csv`.
Matching declaration in Subways.mapcss should match the matching declaration in this file.

Hopefully `[operator=London Overground]` will work out-of-the-box, but I'm not sure.
Also in `generator/osm2type.cpp` there is e.g.
```
          {"network", "Metro de Madrid", [&params] { params.SetRwSubwayType("madrid"); }},
          {"operator", "Metro de Madrid", [&params] { params.SetRwSubwayType("madrid"); }},
```

pastk commented:
Basically you're not adding an icon, but https://github.com/organicmaps/organicmaps/blob/master/docs/STYLES.md#how-to-add-a-new-map-feature--poi-type

pastk commented:
> * Ran: `$ ../omim-build-release/generator_tool --generate_index=true --output="UK_England_Greater London"`

This will just re-build the visibility index in existing map file. It won't add new map features - one has to process the OSM source form scratch for that: https://github.com/organicmaps/organicmaps/blob/master/tools/python/maps_generator/README.md

charlie2clarke commented:
Regenerated the entire UK_Greater_England London map and still no luck. Given that the type isn't being mapped (see below), is the problem in osm2type?
![image](https://github.com/organicmaps/organicmaps/assets/58811769/e59a267f-cbee-49aa-8c26-c4f3017b1fd3)

pastk commented:
> aah, operator is discarded if not useful for the type:

metadata is processed after feature type matching (end of `GetNameAndType()`) and the processing doesn't affect the source tags, so shouldn't be an issue

pastk commented:
> Regenerated the entire UK_Greater_England London map and still no luck. Given that the type isn't being mapped (see below), is the problem in osm2type?

The tags matching happens in `osm2type.cpp::MatchTypes()`. By a quick look it should process whitespace fine, but better debug it.
(looks like you have full files auto-formatting enabled in your IDE, please disable it for OM)

FYI @vng

vng commented on .gitmodules:
```diff
@@ -59,6 +59,3 @@
 	url = https://github.com/harfbuzz/harfbuzz.git
-[submodule "3party/utfcpp"]
-	path = 3party/utfcpp
-	url = https://github.com/nemtrif/utfcpp.git
```
Please, update submodules before committing your changes. This change is not from this PR, definitely.

vng commented on data/mapcss-mapping.csv:
```diff
@@ -1683,3 +1683,4 @@ railway|station|subway|singapore;[railway=station][transport=subway][city=singap
 railway|subway_entrance|singapore;[railway=subway_entrance][city=singapore];;name;int_name;1595;
 railway|station|subway|boston;[railway=station][transport=subway][city=boston],[railway=station][station=subway][city=boston];;name;int_name;1596;
 railway|subway_entrance|boston;[railway=subway_entrance][city=boston];;name;int_name;1597;
+railway|station|overground|london;[railway=station][operator=London Overground];;name;int_name;1598;
```
Well, if it works good (probably should add one test in osm_tyoe_test.cpp) I'm ok with that.

@RedAuburn Can we trust operator=London Overground ?

RedAuburn replied:
hmmm... looks like it's gonna be a bit more complicated :/

this query finds all 113 stations the Overground serves:
```
[out:json][timeout:25];
{{geocodeArea:United Kingdom}}->.searchArea;
// OR
(
  // should match =London Underground;London Overground
  nwr["railway"="station"]["network"~"London Overground"](area.searchArea);
  nwr["railway"="station"]["operator"~"London Overground"](area.searchArea);
);
// print results
out geom;
```

need to search for all nodes where _either_ `network=*` or `operator=*` _contains_ `London Overground` (non-case sensitive). And not just an exact string match as POI can have things like `network=London Underground;London Overground`

pastk replied:
`railway|station|overground|london;[railway=station][operator=London Overground],[railway=station][network=London Overground];` :)

but IDK how do we treat case and multiple values might need a special processing...

charlie2clarke replied:
Would adding another wildcard to signify case insensitivity not be ok? I'd like to help make the changes for how this might be parsed, but I'm having trouble working out how to build and run the generator code by itself (and not driven by tools/python/map_generator) to allow me to debug it.

pastk replied:
IMO all OSM tags matching should be case-insensitive (maybe its the case already?)

RedAuburn replied:
> but IDK how do we treat case and multiple values might need a special processing...
@vng do you know if the generator can handle case & multiple values?

vng replied:
If have doubts, better make tests like in osm_type_test.cpp
